owlkettle/gtk.nim

```diff
@@ -365,7 +366,21 @@ proc g_value_unset*(value: ptr GValue)
 proc g_idle_add_full*(priority: cint, fn: GSourceFunc, data: pointer, notify: GDestroyNotify): cuint
 proc g_timeout_add_full*(priority: cint, interval: cuint, fn: GSourceFunc, data: pointer, notify: GDestroyNotify): cuint
 proc g_strdup*(str: cstring): pointer
+proc g_free*(str: pointer)
+{.pop.}
+
+type OwnedGtkString* = distinct cstring
+
+crossVersionDestructor(cstr, OwnedGtkString):
```

Comment:
It appears I do not necessarily need a =copy hook, because cstring apparently has one and by using cStr.cstring.isNil it implicitly associated the copy-hook from cstring with OwnedGtkString.

Reply:
Are you sure that the `=copy` hook is even created at all? (what happens if you `var a = <some owned gtk string>; var b = a; echo $a; echo $b`) If it uses the `cstring` copy hook, that sounds like it might be bad.

Reply:
I was under the impression the following inim example would save me:
```nim
nim> type A = distinct cstring
nim> let x: A = "lala".A
nim> let y = x
nim> echo x.pointer.repr
000055870C2332B0
nim> echo y.pointer.repr
000055AA38BA02B0
```

Which is basically the scenario we're running.
Turns out that's false because inim works in different ways than I thought. So we're still boned with the damn copy hook.
The entire approach of forward declaring it might work before the pop situation can screw us, but I'm still running into the scenario that to copy a cstring I have to do so without using `=` and how I'm supposed to do that is beyond me.

Reply:
Actually... there is copy behaviour, you get pointer copies, not value copies. Isn't that kinda still okay behaviour?
What is the value of a copy-proc here?

Reply:
Pointer copies don't work in this case, as now the `=destroy` hook is called twice on the same pointer. Which is a classic double free error/mistake.

Reply:
Okay I think I managed to write a copy-hook that works, at least when I compile this minimal example on my end I get no memory corruption, nor segfault, nor recursive nonsense etc., and the pointers point to different memory locations:
```nim
import std/strutils as strutils
{.passl: strutils.strip(gorge("pkg-config --libs gtk4")).}

{.push importc, cdecl.}
proc g_free*(str: pointer)
proc g_strdup*(str: cstring): pointer
proc g_strnfill*(length: uint64, fill_char: char ): pointer
{.pop.}

type OwnedGtkString* = distinct cstring

proc `=destroy`*(cStr: OwnedGtkString)
proc `=copy`*(dest: var OwnedGtkString, source: OwnedGtkString)

proc `isNil`(cStr: OwnedGtkString): bool {.borrow.}

proc `=destroy`*(cStr: OwnedGtkString) =
  echo "Destroy owned gtk string: ", $cstr.cstring
  if cStr.cstring.isNil():
    return
  
  g_free(pointer(cStr))

proc `=copy`*(dest: var OwnedGtkString, source: OwnedGtkString) =
  let areSameObject = pointer(source) == pointer(dest)
  if areSameObject:
    return
  
  `=destroy`(dest)
  wasMoved(dest)
  let copyC: pointer = g_strdup(cstring(source))
  cstring(dest) = cast[cstring](copyC)


let x: OwnedGtkString = cast[cstring](g_strnfill(5, 'x')).OwnedGtkString
let y = x
echo x.pointer.repr
echo y.pointer.repr
```

Reply:
Looks good so far. I would do a `=sink` by following the pattern here: https://nim-lang.org/docs/destructors.html#lifetimeminustracking-hooks-nimeqsink-hook, just to be sure, because of the weird bug we encountered with `distinct` types and hooks.

Reply:
I've done so and pushed it, but I can't really test it.
Well, I'm trying to, but I'm struggling with manipulating the code to do what I want here, at least I can't manage to confirm that the copy hook gets called in the picture.nim example